es-rest-client-1.3/core/src/main/java/com/bazaarvoice/elasticsearch/client/core/util/ExceptionUtil.java

```diff
@@ -0,0 +1,47 @@
+package com.bazaarvoice.elasticsearch.client.core.util;
+
+import org.elasticsearch.index.Index;
+import org.elasticsearch.indices.IndexMissingException;
+
+import java.util.Map;
+import java.util.Objects;
+
+public final class ExceptionUtil {
+
+    private ExceptionUtil() {
+    }
+
+    /**
+     * Gets the appropriate exception from a result map from a RESTful call. The map is
+     * expected to have an entry for 'error' that described the error that occurred.
+     *
+     * Currently only supports special handling for IndexMissingExceptions. If the
+     * cause is an IndexMissingException then that will be returned otherwise the
+     * error message will be wrapped with a RuntimeException and returned.
+     *
+     * @param map non-null result map from RESTful call
+     * @return proper exception; IndexMissingException or RuntimeException with error message
+     *          in the 'error' map entry.
+     */
+    public static RuntimeException getProperException(Map<String, Object> map) {
+
+        RuntimeException ex = new RuntimeException("Encountered error: " + Objects.toString(map));
+
+        // If its an index missing exception then create proper exception to return to caller
+        String errorMessage = Objects.toString(map.get("error"));
+        if (errorMessage.contains("IndexMissingException")) {
+            String missingIndexName = errorMessage;
+            int beginIndex = errorMessage.indexOf("[[");
+            if (beginIndex != -1) {
+                beginIndex += 2; // Length of '[['
+                int endIndex = errorMessage.indexOf("]", beginIndex);
+                if (endIndex != -1) {
+                    missingIndexName = errorMessage.substring(beginIndex, endIndex).trim();
+                }
+            }
```

Comment:
This may be one of those times when Regex actually makes things easier ;)

In the interests of not falsely identifying things as IMEs, we should match as strictly as possible. Based on what I'm seeing, I guess it will be like: `^IndexMissingException\[\[([^\]]+)].*`.

It might be nice to subclass IME so you can add a field containing the original response map. This will help folks who encounter a weird edge case and inspect the exception in the debugger.